src/python/foglamp/device/ingest.py

```diff
@@ -53,62 +50,246 @@ class Ingest(object):
+        loop_ctr = 2*cls._min_queue_flush_seconds
+
+        while True:
+            # Wait for enough items in the queue to fill a batch
+            # for some minimum amount of time
+            for _ in range(loop_ctr):
+                if cls._stop:
+                    break
+                if queue.qsize() >= cls._max_batch_items:
+                    break
+                # _LOGGER.debug('Waiting: Queue index: %s Size: %s',
+                #              queue_index, queue.qsize())
+                await asyncio.sleep(.5)
+
+            # Wait for the first entry
+            try:
+                insert = queue.get_nowait()
+            except asyncio.QueueEmpty:
+                if cls._stop:
+                    break
+
+                # Wait for an item in the queue
+                waiter = asyncio.ensure_future(queue.get())
+                cls._insert_readings_wait_tasks[queue_index] = waiter
+
+                try:
+                    insert = await asyncio.wait_for(waiter, cls._max_idle_db_connection_seconds)
+                except asyncio.CancelledError:
+                    continue
+                except asyncio.TimeoutError:
+                    await cls._close_connection(connection)
+                    connection = None
+                    continue
+                finally:
+                    cls._insert_readings_wait_tasks[queue_index] = None
+
+            inserts = [(insert[0], insert[1], insert[2], json.dumps(insert[3]))]
+            # inserts = [insert]
+
+            for _ in range(1, cls._max_batch_items):
+                if not (len(inserts) % cls._batch_yield_num_items):
+                    await asyncio.sleep(0)
+                try:
+                    insert = queue.get_nowait()
+                except asyncio.QueueEmpty:
+                    break
+                inserts.append((insert[0], insert[1], insert[2], json.dumps(insert[3])))
+                # inserts.append(insert)
+
+            # _LOGGER.debug('Begin insert: Queue index: %s Batch size: %s', queue_index, len(inserts))
+
+            for _ in range(cls._max_insert_attempts):
+                try:
+                    if connection is None:
+                        connection = await asyncpg.connect(database='foglamp')
+                        # Create a temp table for 'copy' command
+                        await connection.execute('create temp table t_readings '
+                                                 'as select asset_code, user_ts, read_key, reading '
+                                                 'from foglamp.readings where 1=0')
+                    else:
+                        await connection.execute('truncate table t_readings')
+
+                    await connection.copy_records_to_table( table_name='t_readings',
+                                                           records=inserts)
+
+                    await connection.execute('insert into foglamp.readings '
```

Comment:
we should pick an item of queue and COPY it. I think this operation create temp table, insert readings, copy to main readings is too expensive.